core/field_checkbox.js

```diff
@@ -92,49 +93,72 @@ Blockly.FieldCheckbox.prototype.init = function() {
       this.fieldGroup_);
   var textNode = document.createTextNode(Blockly.FieldCheckbox.CHECK_CHAR);
   this.checkElement_.appendChild(textNode);
-  this.checkElement_.style.display = this.state_ ? 'block' : 'none';
+  this.checkElement_.style.display = this.value_ ? 'block' : 'none';
+
+  if (this.borderRect_) {
+    this.borderRect_.setAttribute('width',
+        this.size_.width + Blockly.BlockSvg.SEP_SPACE_X);
+  }
 };
 
 /**
- * Return 'TRUE' if the checkbox is checked, 'FALSE' otherwise.
- * @return {string} Current state.
+ * Checkboxes have a constant width.
+ * @private
  */
-Blockly.FieldCheckbox.prototype.getValue = function() {
-  return String(this.state_).toUpperCase();
+Blockly.FieldCheckbox.prototype.render_ = function() {
+  this.size_.width = 5;
+};
+
+/**
+ * Toggle the state of the checkbox on click.
+ * @protected
+ */
+Blockly.FieldCheckbox.prototype.showEditor_ = function() {
+  this.setValue(!this.value_);
+};
+
+/**
+ * Validate the input value to a boolean.
+ * @param {*} newValue The input value.
+ * @return {string} The value converted to a 'TRUE' or 'FALSE' value.
+ * @protected
+ */
+Blockly.FieldCheckbox.prototype.doClassValidation_ = function(newValue) {
+  return this.convertValueToBool_(newValue) ? 'TRUE' : 'FALSE';
 };
 
 /**
- * Set the checkbox to be checked if newBool is 'TRUE' or true,
- * unchecks otherwise.
- * @param {string|boolean} newBool New state.
+ * Store the new value of the field and update the checkElement.
+ * @param {boolean|string} newValue The new boolean value of the field.
+ * @protected
  */
-Blockly.FieldCheckbox.prototype.setValue = function(newBool) {
-  var newState = (typeof newBool == 'string') ?
-      (newBool.toUpperCase() == 'TRUE') : !!newBool;
-  if (this.state_ !== newState) {
-    if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
-      Blockly.Events.fire(new Blockly.Events.BlockChange(
-          this.sourceBlock_, 'field', this.name, this.state_, newState));
-    }
-    this.state_ = newState;
-    if (this.checkElement_) {
-      this.checkElement_.style.display = newState ? 'block' : 'none';
-    }
+Blockly.FieldCheckbox.prototype.doValueUpdate_ = function(newValue) {
+  this.value_ = this.convertValueToBool_(newValue);
+  // Update visual.
+  if (this.checkElement_) {
+    this.checkElement_.style.display = this.value_ ? 'block' : 'none';
```

Comment:
This UI update looks like it should happen in the rerender, with `doValueUpdate` just storing the new value and marking the field as dirty.

Reply:
Actually, I'm seeing this as a pattern across other fields as well.  More work is happening in doValueUpdate than I expect, and less in render, which often just sets the width.

Reply:
Doing stuff inside doValueUpdate_ or doValueInvalid is actually a little bit of an optimization. Setting isDirty to true (and doing work inside render_) means the whole block has to rerender, which is unecessary if the fields size doesn't change.

The reason for setting the width inside render is because setVisible is broken at the moment. It's setting the field's size_.width to zero when that should really be handled inside getSize.

Reply:
Interesting.  Sounds like that's a separate issue, but one worth looking into (avoiding a rerender if the field's size didn't change).  Can you create an issue for it and note which fields will likely need investigation?

Reply:
Can do, do you want everything moved to render then @rachel-fenichel ?

Reply:
Leave it here, and then move it in a followup.

Reply:
Sorry, I think I got myself a bit confused, and I'm unsure of exactly what you're looking for.

I thought I was supposed to remove the optimization, and then put up an issue to re-add the optimization once I could prove it was actually more efficient. So I asked:
> do you want everything moved to render then?  

To make sure I was understanding correctly.

But now it seems like I'm not supposed to remove the optimization? Am I'm still supposed to create an issue to investigate the optimization?

I'm cool with doing whatever! I just wanted to make sure I'm not creating any confusing issues. Sorry I had to bug you again, and thank you for the helps!

Reply:
I think you've made a good case that it's more efficient, and my issue with it is just that the code reads in a confusing way--the meat of rendering the field doesn't always happen in render or a helper called by render.  So I meant 
- you should leave the code as it is right now (correct and efficient).
- we should (separately) add a code path that rerenders only the field and not the block if the field's size doesn't change.
- once that path exists, we should move this code (and similar for other fields) back into the render function or its helpers

Does that make sense?

Reply:
> * you should leave the code as it is right now (correct and efficient).
> * we should (separately) add a code path that rerenders only the field and not the block if the field's size doesn't change.
> * once that path exists, we should move this code (and similar for other fields) back into the render function or its helpers
> 
> Does that make sense?

Makes sense! I put up #2490 with both proposed solutions.